sycl/source/detail/event_impl.cpp

```diff
@@ -93,8 +101,62 @@ event_impl::event_impl(QueueImplPtr Queue) : MQueue(Queue) {
   }
 }
 
+void *event_impl::instrumentationProlog(string_class &Name, int32_t StreamID,
+                                        uint64_t &IId) const {
+  void *TraceEvent = nullptr;
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  static uint64_t instance_id = 1;
+  xpti::trace_event_data_t *WaitEvent = nullptr;
+  if (xptiTraceEnabled()) {
+    // Create a string with the event address so it
+    // can be associated with other debug data
+    xpti::utils::StringHelper SH;
+    Name = SH.nameWithAddress<RT::PiEvent>("event.wait", MEvent);
+
+    // We can emit the wait associated with the graph if the
+    // event does not have a command object or associated with
+    // the command object, if it exists
+    if (MCommand) {
+      Command *Cmd = (Command *)MCommand;
+      WaitEvent = Cmd->MTraceEvent ? static_cast<xpti_td *>(Cmd->MTraceEvent)
+                                   : GSYCLGraphEvent;
+    } else
+      WaitEvent = GSYCLGraphEvent;
+
+    // Record the current instance ID for use by Epilog
+    IId = instance_id;
+    xptiNotifySubscribers(StreamID, xpti::trace_wait_begin, nullptr, WaitEvent,
+                          instance_id++,
```

Comment:
Races can happen here, increment operation is not thread safe.

Reply:
@romanovvlad Good catch - uses atomic now.